lib/seattleflu/db/cli/command/etl/consensus_genome.py

```diff
@@ -0,0 +1,363 @@
+from . import etl
+
+
+LOG = logging.getLogger(__name__)
+
+
+# This revision number is stored in the processing_log of each consensus genome
+# record when the consensus genome record is successfully processed by this ETL
+# routine. The routine finds new-to-it records to process by looking for
+# consensus genome records lacking this revision number in their log.  If a
+# change to the ETL routine necessitates re-processing all consensus genome
+# records, this revision number should be incremented.
+REVISION = 1
+
+
+@etl.command("consensus-genome", help = __doc__)
+
+@click.option("--dry-run", "action",
+    help        = "Only go through the motions of changing the database (default)",
+    flag_value  = "rollback",
+    default     = True)
+
+@click.option("--prompt", "action",
+    help        = "Ask if changes to the database should be saved",
+    flag_value  = "prompt")
+
+@click.option("--commit", "action",
+    help        = "Save changes to the database",
+    flag_value  = "commit")
+
+def etl_consensus_genome(*, action: str):
+    LOG.debug(f"Starting the consensus genome ETL routine, revision {REVISION}")
+
+    db = DatabaseSession()
+
+    # Fetch and iterate over consensus genome records that aren't processed
+    #
+    # Rows we fetch are locked for update so that two instances of this
+    # command don't try to process the same consensus genome records.
+    LOG.debug("Fetching unprocessed consensus genome records")
+
+    consensus_genome = db.cursor("consensus genome")
+    consensus_genome.execute("""
+        select consensus_genome_id as id, document, received
+          from receiving.consensus_genome
+         where not processing_log @> %s
+         order by id
+           for update
+        """, (Json([{ "revision": REVISION }]),))
+
+    processed_without_error = None
+
+    try:
+        for record in consensus_genome:
+            with db.savepoint(f"consensus genome record {record.id}"):
+                LOG.info(f"Processing consensus genome record {record.id}")
+
+                # Verify sample identifier is in the database
+                sample = find_sample(db, record.document)
+
+                # Most of the time we expect to see existing sequence read sets,
+                # but we also want to update the details log. However, the
+                # only unique constraint on the sequence_read_set table is
+                # defined within a trigger function, as Postgres does not allow
+                # unique constraints on array columns. Therefore, perform a
+                # find-or-create followed by an update to the details column to
+                # avoid conflict.
+                sequence_read_set = find_or_create_sequence_read_set(db,
+                    record.document, sample)
+
+                mark_received(db,
+                    sequence_read_set.id,
+                    {   "received": record.received,
+                        "consensus_genome_id": record.id    })
```

Comment:
"consensus_genome_id" might be a little misleading here? One might think it's the `genome_id` is `warehouse.genome`

Reply:
I thought it might be useful to tie in the consensus genome id from `receiving.consensus_genome`. Do you have a suggestion for a better name, or should this idea be abandoned?